core/chains/evm/txm/txm.go

```diff
@@ -0,0 +1,443 @@
+				t.lggr.Warnf("Reached transaction limit. LocalNonce: %d, PendingNonce %d, unconfirmedCount: %d",
+					nonce, pendingNonce, unconfirmedCount)
+				return true, nil
+			}
+		}
+
+		tx, err := t.txStore.UpdateUnstartedTransactionWithNonce(ctx, address, t.getNonce(address))
+		if err != nil {
+			return false, err
+		}
+		if tx == nil {
+			return false, nil
+		}
+		nonce := t.getNonce(address)
+		tx.Nonce = &nonce
+		t.setNonce(address, *tx.Nonce+1)
+		tx.State = types.TxUnconfirmed
+
+		if err := t.createAndSendAttempt(ctx, tx, address); err != nil {
+			return true, err
+		}
+	}
+}
+
+func (t *Txm) createAndSendAttempt(ctx context.Context, tx *types.Transaction, address common.Address) error {
+	attempt, err := t.attemptBuilder.NewAttempt(ctx, t.lggr, tx, t.config.EIP1559)
+	if err != nil {
+		return err
+	}
+
+	if tx.Nonce == nil {
+		return fmt.Errorf("nonce for txID: %v is empty", tx.ID)
+	}
+	if err = t.txStore.AppendAttemptToTransaction(ctx, *tx.Nonce, address, attempt); err != nil {
+		return err
+	}
+
+	return t.sendTransactionWithError(ctx, tx, attempt, address)
+}
+
+func (t *Txm) sendTransactionWithError(ctx context.Context, tx *types.Transaction, attempt *types.Attempt, address common.Address) (err error) {
+	if tx.Nonce == nil {
+		return fmt.Errorf("nonce for txID: %v is empty", tx.ID)
+	}
+	start := time.Now()
+	txErr := t.client.SendTransaction(ctx, tx, attempt)
+	tx.AttemptCount++
+	t.lggr.Infow("Broadcasted attempt", "tx", tx.PrettyPrint(), "attempt", attempt.PrettyPrint(), "duration", time.Since(start), "txErr: ", txErr)
+	if txErr != nil && t.errorHandler != nil {
+		if err = t.errorHandler.HandleError(tx, txErr, t.attemptBuilder, t.client, t.txStore, t.setNonce, false); err != nil {
+			return
+		}
+	} else if txErr != nil {
+		pendingNonce, err := t.client.PendingNonceAt(ctx, address)
+		if err != nil {
+			return err
+		}
+		if pendingNonce <= *tx.Nonce {
+			t.lggr.Debugf("Pending nonce for txID: %v didn't increase. PendingNonce: %d, TxNonce: %d", tx.ID, pendingNonce, tx.Nonce)
+			return nil
+		}
+	}
+
+	promNumBroadcastedTxs.WithLabelValues(t.chainID.String()).Add(float64(1))
+	return t.txStore.UpdateTransactionBroadcast(ctx, attempt.TxID, *tx.Nonce, attempt.Hash, address)
+}
+
+func (t *Txm) backfillTransactions(ctx context.Context, address common.Address) (bool, error) {
+	latestNonce, err := t.client.NonceAt(ctx, address, nil)
+	if err != nil {
+		return false, err
+	}
+
+	confirmedTransactions, unconfirmedTransactionIDs, err := t.txStore.MarkTransactionsConfirmed(ctx, latestNonce, address)
+	if err != nil {
+		return false, err
+	}
+	if len(confirmedTransactions) > 0 || len(unconfirmedTransactionIDs) > 0 {
+		promNumConfirmedTxs.WithLabelValues(t.chainID.String()).Add(float64(len(confirmedTransactions)))
+		confirmedTransactionIDs := extractMetrics(confirmedTransactions, t.chainID)
+		t.lggr.Infof("Confirmed transaction IDs: %v . Re-orged transaction IDs: %v", confirmedTransactionIDs, unconfirmedTransactionIDs)
```

Comment:
Since we're printing IDs here, shouldn't this log be Debug level?

Reply:
Also are IDs really helpful here for debugging since its just an internal sequence? Wouldn't the current nonce be more helpful and maybe the count of transactions confirmed/re-org'd?

Reply:
So the general idea for TXM's logs is that you don't depend on a specific log to view the entire information, but you read them all together. This way we avoid spamming the logs with repeated information. Another point, is that I've used txID as a key for most of the logs, so if something goes wrong we can search the logs for that ID and get the whole story. For example, with just the info level someone can look at the message `Broadcasted attempt...txID` and `Confirmed transaction IDs` and see when the transaction was broadcasted and when it was confirmed. I could see how adding nonce could improve this, but I think ID should be the general source of truth. Keep in mind we can easily derive nonce from an ID but not the other way around.